src/basic/in-addr-util.c

```diff
@@ -581,9 +581,11 @@ void in_addr_data_hash_func(const void *p, struct siphash *state) {
 
 int in_addr_data_compare_func(const void *a, const void *b) {
         const struct in_addr_data *x = a, *y = b;
+        int r;
 
-        if (x->family != y->family)
-                return x->family - y->family;
+        r = CMP(x->family, y->family);
+        if (r != 0)
+                return r;
```

Comment:
What about
```C
if (x->family != y->family)
     return CMP(x->family, y->family)
```
?
This is slightly shorter and doesn't require a temporary variable. The compiler should optimize the repeated check away without any trouble.

Reply:
I did think of that... But in the end I saw some compare_func's that use strcmp assigning to an `r` variable, which made me prefer to use the `if (r != 0)` form since it could match use of those functions too. I ended up used that form everywhere, for consistency.

See @yuwata's suggestion on `CMP_return()` and my follow up on `return_if_nonzero()` (which could also be used for strcmp and memcmp), I think that would be a more general way to solve this. WDYT?

Reply:
so far we don't have a return_xyz() family of macros, but we do have assert_return_xyz() which might return internally. I am not really sure what the best approach here is, and if we should add generic return_if_nonzero(), since I usually don't like such "hidden" jumps too much... (I wasn't even sure about the assert_return_xyz() case, but I think it's more well defined, since it usually appears as part of the header of a function only).

Hence, maybe we should call it `RETURN_CMP_VERDICT()` or so, which hopefully signals that it only does something when a verdict is found (i.e. when the predicate is non-zero), but we make it specific enough for the comparison case so that we don't end up with it all over the place in other areas...

```c
#define RETURN_CMP_VERDICT(x) \
        ({  \
                typeof(x) y = x; \
                if (y != 0) \
                        return y; \
        )}
```

or something like this?

You could then call it like this:

```c
RETURN_CMP_VERDICT(CMP(foo, bar));
RETURN_CMP_VERDICT(strcmp(x, y));
…
```

Reply:
How about we start using `||` for this?

```c
int in_addr_data_compare_func(const void *a, const void *b) {
        const struct in_addr_data *x = a, *y = b;
        return (CMP(x->family, y->family) ||
                memcmp(&x->address, &y->address, FAMILY_ADDRESS_SIZE(x->family)));
}
```

And:

```c
static int routing_policy_rule_compare_func(const void *_a, const void *_b) {
…
        case AF_INET6:
                return (CMP(a->from_prefixlen, b->from_prefixlen) ||
                        CMP(a->to_prefixlen, b->to_prefixlen) ||
                        CMP(a->tos, b->tos) ||
                        CMP(a->fwmask, b->fwmask) ||
                        CMP(a->table, b->table) ||
                        strcmp_ptr(a->iif, b->iif) ||
                        strcmp_ptr(a->oif, b->oif) ||
                        memcmp(&a->from, &b->from, FAMILY_ADDRESS_SIZE(a->family)) ||
                        memcmp(&a->to, &b->to, FAMILY_ADDRESS_SIZE(a->family)));
```

Only thing here is that you need to understand how `||` inteacts with 0 and non-zero integers... But considering we have variables that can take true, false or -1, that's not much different :-)

This accomplishes terseness on the code, without having to introduce macros that result in breaking control flow (returning sometimes, not always.)

We wouldn't solve all cases here, but many of them would work.

In any case, I'm leaning towards not doing this in the same PR, but just doing CMP() for now (maybe even in the cases where this increases the total number of lines from 2 to 3) and then we can revisit this in a follow up.

Reply:
Hmm, not following? The || operator always resolves to 0 or 1, no? never to -1? what am i missing? when sorting we need the order between two values, and hence || can't work, or I am missing something?

Reply:
Hmmm, my bad. I thought it worked the same as in Perl or Python, where it returns the first non-zero value...

Though maybe a variadic C macro that would do just that might be a good approach here, perhaps?